File: Week_01/cpp/src/_143_ReorderList.cpp
#include <iostream>
using namespace std;
struct ListNode
{
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};
class Solution
{
public:
    // 没有想到先逆序，然后再合并。。
    void reorderList(ListNode *head)
    {
        if (!head)
            return;
        // 找到中间节点
        auto p = midOf(head);
        // 将后半段链表反转
        auto q = reverse(p->next);
        p->next = NULL;
        // 合并两个链表
        merge(head, q);
    }
    void merge(ListNode *p, ListNode *q)
    {
        while (q)
        {
            auto t = p->next;
            p->next = q;
            q = q->next;
            p->next->next = t;
            p = t;
        }
    }
    ListNode *reverse(ListNode *p)
    {
        ListNode *head = NULL;
        while (p)
        {
            auto q = p->next;
            p->next = head;
            head = p;
            p = q;
        }
        return head;
    }
    // 节点数为奇数则返回中间节点，否则返回前半段最后一个节点
    ListNode *midOf(ListNode *p)
    {
        auto q = p;
        while (p->next && p->next->next)
        {
            p = p->next->next;
            q = q->next;
        }
        return q;
    }
};

File: Week_01/kotlin/src/_143_ReorderList.kt
class _143_ {
    /**
     * Example:
     * var li = ListNode(5)
     * var v = li.`val`
     * Definition for singly-linked list.
     * class ListNode(var `val`: Int) {
     *     var next: ListNode? = null
     * }
     */
    class Solution {
        fun reorderList(head: ListNode?): Unit {
            head ?: return
            // 1.找到中间节点
            fun midOf(h: ListNode?): ListNode? {
                var p = h
                var q = h
                while (p?.next != null && p?.next?.next != null) {
                    p = p?.next?.next
                    q = q?.next
                }
                return q
            }

            val m = midOf(head)

            // 2.反转后半段
            fun reverse(p: ListNode?): ListNode? {
                var h: ListNode? = null
                var q = p
                while (q != null) {
                    val t = q?.next
                    q?.next = h
                    h = q
                    q = t
                }
                return h
            }

            val r = reverse(m?.next)
            m?.next = null
            // 3.合并两段链表
            fun merge(p: ListNode?, q: ListNode?): Unit {
                var l = p
                var r = q
                while (r != null) {
                    val t = r?.next
                    r?.next = l?.next
                    l?.next = r
                    l = r?.next
                    r = t
                }
            }
            merge(head, r)
        }
    }
}